RubyHook.cpp

```diff
@@ -81,6 +144,41 @@ RubyHook::RubyHook(const mesos::Parameters& parameters)
   }
 }
 
+// mesos::Hook callback on slaveExecutorEnvironmentDecorator
+Result<mesos::Environment> RubyHook::slaveExecutorEnvironmentDecorator(
+  const mesos::ExecutorInfo& executorInfo)
+{
+  synchronized (mutex) { // Ruby VM is not reentrant
+    if (ruby.is_callback_defined(SlaveExecutorEnvironmentDecorator)) {
+      // call Ruby in a protect env to avoid exception leakage
+      int state = 0;
+      VALUE executorInfoValue = wrapExecutorInfo(executorInfo); // map ExecutorInfo into strings kv-pair hash
+      VALUE result = rb_protect(::slaveExecutorEnvironmentDecorator_wrapper, executorInfoValue, &state);
+      if (state != 0) {
+        return Error(ruby.handle_exception());
+      }
+
+      // fill in the Env from Ruby from return value; expect same structure as input
+      if (hashNotNullOrEmpty(result)) {
+        VALUE ruby_cmd = rb_hash_lookup(result, rb_str_new_cstr("command"));
+        if (hashNotNullOrEmpty(ruby_cmd)) {
+          VALUE ruby_env = rb_hash_lookup(ruby_cmd, rb_str_new_cstr("environment"));
+          if (hashNotNullOrEmpty(ruby_env)) {
+            mesos::Environment env;
```

Comment:
painful ... OPT: 
hash_lookup( VALUE hash, ... )
https://stackoverflow.com/questions/1657883/variable-number-of-arguments-in-c

Reply:
I'm not comfortable enough to write that :)